src/hotspot/share/runtime/interfaceSupport.inline.hpp

```diff
@@ -76,59 +76,53 @@ class InterfaceSupport: AllStatic {
-    _thread = thread;
-    assert(thread != NULL, "must be active Java thread");
-    assert(thread == Thread::current(), "must be current thread");
-  }
 
-  // Change threadstate in a manner, so safepoint can detect changes.
-  // Time-critical: called on exit from every runtime routine
-  static inline void transition(JavaThread *thread, JavaThreadState from, JavaThreadState to) {
-    assert(from != _thread_in_Java, "use transition_from_java");
-    assert(from != _thread_in_native, "use transition_from_native");
-    assert((from & 1) == 0 && (to & 1) == 0, "odd numbers are transitions states");
-    assert(thread->thread_state() == from, "coming from wrong thread state");
-
-    // Check NoSafepointVerifier
-    // This also clears unhandled oops if CheckUnhandledOops is used.
+ private:
+  static inline void transition_and_process(JavaThread *thread, JavaThreadState to, bool check_asyncs) {
+    // Check NoSafepointVerifier. This also clears unhandled oops if CheckUnhandledOops is used.
     thread->check_possible_safepoint();
 
-    // Change to transition state and ensure it is seen by the VM thread.
-    thread->set_thread_state_fence((JavaThreadState)(from + 1));
-
-    SafepointMechanism::process_if_requested(thread);
+    thread->set_thread_state_fence(_thread_in_vm);
+    SafepointMechanism::process_if_requested_with_exit_check(thread, check_asyncs);
     thread->set_thread_state(to);
   }
 
-  // Same as above, but assumes from = _thread_in_Java. This is simpler, since we
-  // never block on entry to the VM. This will break the code, since e.g. preserve arguments
-  // have not been setup.
+ public:
+  ThreadStateTransition(JavaThread *thread) : _thread(thread) {
+    assert(thread != NULL, "must be active Java thread");
+    assert(thread == Thread::current(), "must be current thread");
+  }
+
   static inline void transition_from_java(JavaThread *thread, JavaThreadState to) {
     assert(thread->thread_state() == _thread_in_Java, "coming from wrong thread state");
+    assert(to == _thread_in_vm || to == _thread_in_native, "invalid transition");
     thread->set_thread_state(to);
   }
 
-  static inline void transition_from_native(JavaThread *thread, JavaThreadState to) {
-    assert((to & 1) == 0, "odd numbers are transitions states");
+  // We never install asynchronous exceptions when coming (back) in to the runtime
+  // from native code because the runtime is not set up to handle exceptions floating
+  // around at arbitrary points.
+  static inline void transition_from_native(JavaThread *thread, JavaThreadState to, bool check_asyncs = true) {
     assert(thread->thread_state() == _thread_in_native, "coming from wrong thread state");
-    assert(!thread->has_last_Java_frame() || thread->frame_anchor()->walkable(), "Unwalkable stack in native->vm transition");
-
-    // Change to transition state and ensure it is seen by the VM thread.
-    thread->set_thread_state_fence(_thread_in_native_trans);
```

Comment:
Why do we no longer need _thread_in_native_trans? If it is truly not needed here then it seems to no longer be used?

Reply:
We just need to transition to an unsafe state before polling to avoid the issue of a safepoint/handshake starting right after we checked but before we transition to the unsafe state. Since we are in the vm we can use _thread_in_vm instead which is already unsafe. We still use the _thread_in_native_trans state on the native wrappers when transitioning out of native back to Java, but I will try to address that in another RFE.

Reply:
I kind of liked the trans states as indicators of unsafe states as it made the logic consistent and simple:
- advance to current state + 1 (trans state)
- check for safepoint
- advance to final state

and you didn't have to think too much about what the initial state was.

But I can see that logically any unsafe state works.

Reply:
Right, but whenever we are in safepoint/handshake code we are already in vm, that's why I actually think directly setting _thread_in_vm here makes more sense than having additional states just to make the poll. In fact using transitional states(along with the code in handle_polling_page_exception) forces us to need yet another transition wrapper for handshakes(ThreadInVMForHandshake) just to switch to vm and back. 
Reading now the description of transition states in globalDefinitions.hpp I see they were more useful back then with the old safepoint logic:

"These extra states makes it possible for the safepoint code to handle certain thread_states without having to suspend the thread - making the safepoint code faster."

In SafepointSynchronize::block() there was a faster path for _thread_in_native_trans and _thread_blocked_trans (https://github.com/openjdk/jdk/blame/66aa45649ad36ed41c0241ded767d465248eda3c/src/hotspot/share/runtime/safepoint.cpp#L816).

Reply:
Yes things have changed a lot :)